evolver/calibration/interface.py

```diff
@@ -109,16 +113,25 @@ class Calibrator(BaseInterface):
     A modular layer for encapsulating the calibration procedure and data transformations.
     """
 
-    # Calibration state, this is where the calibration data is stored - TODO refactor to use CalibrationData instead.
-    class state(_BaseConfig):
-        status: str | None = "not calibrated"
-
     class Config(Transformer.Config):
         input_transformer: ConfigDescriptor | Transformer | None = None
         output_transformer: ConfigDescriptor | Transformer | None = None
         calibration_file: str | None = None
 
-    class CalibrationData(Transformer.Config): ...
+    class CalibrationData(Transformer.Config):
+        """Stores calibration data, including the measured_data in the CalibrationProcedure.
+
+        While the CalibrationProcedure attached to a Calibrator instance may hold state information, it will not
+        be persisted between sessions. This class is intended to store the state information in a file that can be
+        loaded and saved as needed.
+
+        The CalibrationProcedure must explicitly include an action to "save" CalibrationProcedure state to the
+        Calibrator's CalibrationData.
+        """
+
+        def save_calibration_procedure_state(self, calibration_procedure_state: dict[str, Any]):
+            self.calibration_procedure_state = calibration_procedure_state
+            self.save()
```

Comment:
Shouldn't this just be ``self.save()`` from the caller, why do we need a separate func for this?

E.g., if the caller is able to call ``save_calibration_procedure_state()`` then they already have an instance of ``CalibrationData``, which is a model, and therefor it's state should have been populated upon instantiation and shouldn't just be ``self.calibration_procedure_state`` but all the fields that _are_ ``calibration_procedure_state``, right?

Reply:
As I understand it - Calibrator has a CalibrationProcedure that has state. Sometimes, CalibrationProcedure "pushes" it's state to Calibrator.CalibrationData for it to be saved to file...


here's what this looks like in the latest revision - note the calibration_procedure_stateparam) on  CalibrationData.Config.save

Is there a nicer way to do this?

```
def save(
            self,
            file_path: Path = None,
            encoding: str | None = None,
            calibration_procedure_state: ProcedureStateModel = None,
        ):
            """
            Save the calibration procedure state to a file. This should be triggered by a calibration procedure action
            when the calibration procedure is complete, or whenever it makes sense for example after each action, or
            after a set of complex or laborious actions, when loosing the procedure state would be detrimental.

            Args:
                calibration_procedure_state (ProcedureStateModel): The state data from the calibration procedure to be saved.
                file_path (Path, optional): The path to save the file. Defaults to a generated name based on `name` and `created`.
                encoding (str, optional): The file encoding. Defaults to None.
            """
            self.calibration_procedure_state = calibration_procedure_state
            if file_path is None:
                file_path = Path(f"{self.name}_{self.created.strftime(settings.DATETIME_PATH_FORMAT)}").with_suffix(
                    ".yml"
                )
            return super().save(file_path=self.dir / file_path, encoding=encoding)
            ```